vmicore/src/lib/vmi/RegisterEventSupervisor.h

```diff
@@ -0,0 +1,76 @@
+#ifndef VMICORE_REGISTEREVENTSUPERVISOR_H
+#define VMICORE_REGISTEREVENTSUPERVISOR_H
+
+#include "../io/ILogging.h"
+#include "Event.h"
+#include "LibvmiInterface.h"
+#include <functional>
+#include <memory>
+#include <stdexcept>
+
+namespace VmiCore
+{
+    class IRegisterEventSupervisor
+    {
+      public:
+        virtual ~IRegisterEventSupervisor() = default;
+
+        virtual void teardown() = 0;
+
+        virtual void initializeDtbMonitoring() = 0;
+
+        virtual void setContextSwitchCallback(const std::function<void(vmi_event_t*)>& eventCallback) = 0;
+
+        template <class T>
+        static std::function<void(vmi_event_t*)> createContextSwitchCallback(std::shared_ptr<T> callbackObjectShared,
+                                                                             void (T::*callbackFunction)(vmi_event_t*))
+        {
+            return [callbackObject = std::weak_ptr<T>(callbackObjectShared),
+                    callbackFunction](vmi_event_t* contextSwitchEvent)
```

Comment:
This is old code from the pre C++17 era. Nowadays, it's possible to use `weak_from_this()`, so you can simply write:
```suggestion
        static std::function<void(vmi_event_t*)> createContextSwitchCallback(std::weak_ptr<T> callbackObject,
                                                                             void (T::*callbackFunction)(vmi_event_t*))
        {
            return [callbackObject, callbackFunction](vmi_event_t* contextSwitchEvent)
```